### PR TITLE
[SC] Value transfer rollback

### PR DESCRIPTION
**Pull request by hyunsooda**
## Proposed changes

Ideally, the value transfer from one chain to the other chain is initiated by end-users with enough balance to make a value transfer transaction. The current implementation, however, had been vulnerable to public usage. First of all, value transfer can be failed for a series of reasons of the other chain’s state, which results in the loss of the sender’s KLAY or tokens without return. Thus, we considered a rollback feature to refund for failed value transfers. Second, the value transfer recovery (VTR) was motivated to resend unhandled value transfer requests just in case of the absence of a connected bridge node. However, its implementation simply resends if a corresponding nonce is not marked (handled), including all unexecuted and failed cases. For the latter, the failed transactions must not be resent, reason its failure, and decide whether to refund or not. Eventually, this change implements value transfer rollback.

Changes are:
1. Added three database write and read operations for all the pair of request value transfer tx and event with handle value transfer tx and event, failed handle value transfer tx and event, refund event and tx. Also, read operations are available with APIs.
2. Implemented rollback. All the failed transactions are refunded. The rollback implementation is based on operator's message integrity and event integrity of the bridge contract.
3. Only unexecuted handle value transfer txs are resent by VTR.
4. As operators pay the value transfer services on each chain, the service fee to be paid to operators is reasonable. If the fee is costly, the service may not be attractive. The operator will make a loss if the fee is highly cheap or zero. This change also provides an API for the suggestion of the least fee to not have a deficit.
5. Operators can withdraw KLAY or tokens that a bridge contract holds by the value transfer request. The amount of KLAY or tokens is equally divided by each operator. Users can make a refund call if the reserved period is passed.
6. Refined the log of summarized transfer.
    - Before: `request=, handle=, lowerHandle=, pending=`
    - Now: `request=, handled=, numberOfFailed=, numberOfRefunds=, lowerHandle=, pending=`

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hyunsooda** [comment]
cc. @2dvorak

**Comment by hyunsooda** [comment]
@hyunsooda, TODO: doc update

File: node/sc/vt_reasoning_test.go [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+func cleanup(t *testing.T, sim *backends.SimulatedBackend, dataDir string, exit chan struct{}) {
+	exit <- struct{}{}
+	<-exit
+	sim.Close()
+	if err := os.RemoveAll(dataDir); err != nil {
+		t.Fatalf("fail to delete file %v", err)
+	}
+}
+
+func TestKLAYReasoing(t *testing.T) {
```
Comment (hqjang-pepper) [inline node/sc/vt_reasoning_test.go:293]:
```suggestion
func TestKLAYReasoning(t *testing.T) {
```

File: node/sc/sub_event_handler.go [hunk 2]
```diff
@@ -85,21 +85,42 @@ func (cce *ChildChainEventHandler) ProcessRequestEvent(ev IRequestValueTransferE
...
+	return reqBridgeInfo.removeRefundLedger(ev.HandleNonce)
+}
+
+func (cce *ChildChainEventHandler) ProcessRefundEvent(ev *RefundEvent) error {
+	refundBridgeInfo, ok := cce.subbridge.bridgeManager.GetBridgeInfo(ev.Raw.Address)
+	if !ok {
+		return errors.New("there is no bridge")
+	}
+	refundBridgeInfo.SetRefundEvents()
+	logger.Trace("[SC][Refund] Refund event received", "reqNonce", ev.RequestNonce, "sender", ev.Sender, "value", ev.Value)
```
[the remark this answers is not in the document]
Comment (hyunsooda) [reply node/sc/sub_event_handler.go]: Thank you. Revised.

File: node/sc/vt_reasoning_test.go (hunk 1 above)
Comment (hyunsooda) [reply node/sc/vt_reasoning_test.go]: Thank you. Revised.

File: contracts/bridge/BridgeTransfer.sol [hunk 3]
```diff
@@ -31,8 +31,17 @@ contract BridgeTransfer is BridgeHandledRequests, BridgeFee, BridgeOperator {
...
     uint64 public upperHandleNonce; // a maximum nonce of the counterpart bridge's value transfer request that is handled.
     uint64 public recoveryBlockNumber = 1; // the block number that recovery start to filter log from.
+    uint64 public failedHandles; // a count of failed handle value transfer
+    uint64 public nRefunds; // a count of refunds
     mapping(uint64 => uint64) public handleNoncesToBlockNums;  // <request nonce> => <request blockNum>
 
+    mapping(uint64 => address payable) public refundAddrMap; // <nonce of request, sender>
+    mapping(uint64 => uint256) public refundValueMap; // <nonce of request, value>
+    mapping(uint64 => TokenType) public refundTokenType; // <nonce of request, token type>
+    mapping(uint64 => uint) public refundTimestampMap; // <nonce of request, token type>
```
Comment (hqjang-pepper) [inline contracts/bridge/BridgeTransfer.sol:41]:
```suggestion
    mapping(uint64 => uint) public refundTimestampMap; // <nonce of request, timestamp>
```

**Comment by hyunsooda** [comment]
pending

**Pull request closed without merging**